### ISSUE TITLE
Partager un compte-rendu

### ISSUE THREAD

**Issue by PabloPrudhommeau** [Hypertopic/Cassandre#41]
All the listed tools will be available from the memo page

The user can choose to let his memo public or private, in case it's private he can : 
- Share to user(s) or group of user
- Share by date
- Share by memo's object (minutes, observation...)
- Take a look at the shared memos

**Comment by supertinou** [comment]
Well, thanks @PabloPrudhommeau for this preliminary work on this feature :+1: 
@mchidaine, @valentin-bonino and I have few questions about it. Maybe we have some lack of information on the feature #40.

What is the aim of the feature "Share by date" , "share by memo's object" ? What is the value of this feature ?

Possible scenarios :  sharing with others students, sharing with interviewee and sharing with teachers

The few really usefull feature we are thinking about are 
- be able to share a memo (for instance, some memos don't need to be shared because they are private) 
-  be able to share some memo to someone who do not have any account (for instance the  interviewee) and share a unique link to access this memo
- Allow to show which memo are shared and with whom.

Few questions : 
- Is usefull to think about a way to share the memo  with different access rights, for instance:  specify for each memo and user different reading/writing access ?

**Comment by Slaals** [comment]
According to @christophe-lejeune put a date to a memo is an important point. As memo is always dated, we think that sharing by date could be a useful tool, for example a student wants to show to his/her teacher the memo he/she wrote at a particular time.

Sorry, we didn't choose the right word, we should say "type" instead of "object". As you know, a memo can have different types, theorical report for instance. We follow the same logic that you've seen above.

So, to summarize, the concret value of those features is to help users to share their memos effectively.

@supertinou We agree with your 3 features, not much to say.

For your last question, we think that specifics access to memos are essential.

**Comment by bj0rge** [comment]
Thanks @Slaals for the precisions.

@supertinou, we tried to answer to your questions and insert all the feature you suggest into a scenario with @PabloPrudhommeau and @yeahger :
According to [this comment](https://github.com/Hypertopic/Cassandre/issues/40#issuecomment-89241918) concerning adding a tag to associate a piece of text to a specific keyword during writing a memo, we think about the ability to share a part of a memo without having to copy & paste only some parts of the text, but automatically truncated thanks to the tags we have chosen during the writing.

![scenar](https://cloud.githubusercontent.com/assets/3996102/6981023/7f503832-d9fb-11e4-8fed-73a1e8e70fc3.jpg)

**Comment by christophe-lejeune** [comment]
Because users share memos for different reasons, the way it is done may be different too. 
* Every team member of a project has access to all memos of the project (I think)
* However, given operational memos are used for coordination, they fill a specific role. They should be shown as  to do lists to be completed to any member (of the team) when (s)he connects to the platform.
* Sharing with a colleague, another scientist or a teacher requires a more selective action : the user selects some memos to be shared, that helps the colleague to understand the whole research process (if available, the last version of the analytical storyline is always shared in that case). All the memos do not become available. However it could be interesting to provide a map of the whole project (showing the relations between all memos; these relations are provided by the memo's "grounding" field - see #40 ). 
* Sharing with interviewees is different : mainly two kinds of memos could be shared : the last version of the descriptive storyline of actors' experience and the interview transcript itself (which is a field note).

**Comment by EdwardNjango** [comment]
Bonjour, 
Nous vous présentons ci dessous deux maquettes concernant la fonctionnalité. 
La première concerne le partage en mode privé et la seconde le partage en mode publique.

Capture 1:
![capture d ecran 2015-04-09 a 07 50 14](https://cloud.githubusercontent.com/assets/10283167/7061104/23b93e24-de8d-11e4-9e92-3a494eb113f7.png)

Capture 2: 
![capture d ecran 2015-04-09 a 07 52 06](https://cloud.githubusercontent.com/assets/10283167/7061125/8d1bdf02-de8d-11e4-8276-985a86fbe320.png)

**Comment by christophe-lejeune** [comment]
First, let's review the default situation. 
* All the members involved in a project can **read** the memos of the project's diary, even if they did not author it.
* Only the member that authors a memo is allowed to **modify** it. 
* A memo can have more than one author (#40). In that case, all the authors are allowed to edit it. 

Access to one selected memo can be restricted:
* A member can decide that one memo is private. In that case, (s)he is the only person allowed to read it.

Now, let's review how a memo can be shared with someone who is **not** member of the project. Sharing means allowing to **read** a memo, not to modify it. 
* In selected cases (teacher, colleague, mate), this person is a user of the platform, (s)he owns an id and a password. 
* In other cases (customer, lay person, actor), this person is not a user of the platform. These cases raise questions of how the user is identified (email,...) and how the access will be granted (private URL, password...). 

Sharing memos implies other features, such as (#46) monitoring if (and when) a memo is read and (#45) commenting the shared memo.

**Comment by valentinlefevre** [comment]
Après avoir échangé avec @christophe-lejeune , hier matin,et @benel , ce matin,  il semble que les maquettes ci-dessus ne répondent pas complètement aux besoins exprimés. 

Nous vous proposons donc une solution différente qui inclue une refonte de l'organisation des documents composant les projets sur le modèle de GoogleDrive. A la création d'un projet, le créateur aura la possibilité d'inviter un certain nombre de collaborateurs qui se verront attribuer un journal de bord. Ces journaux de bord seraient automatiquement paramétrés comme étant accessibles en écriture et lecture à leur propriétaire, et en lecture et commentaire pour les collaborateurs, ce paramétrage serait modifiable par le propriétaire du journal et il en va de même pour chaque mémo. Il serait bien entendu possible d'envoyer un lien offrant une accessibilité en lecture et commentaire à une personne extérieure au projet/Journal de bord/Mémo (professionnelle ou autre). Cela nous permet d'allier le côté "privé" d'un journal de bord au partage et à la collaboration nécessaires à la bonne marche d'un projet.

Concernant le rendu, afin de ne pas réinventer la roue de suite nous avons réalisé des captures d'écran de GoogleDrive afin d'appuyer notre proposition par des exemples visuels (des maquettes balsamiq viendront par la suite si cela convient.

- Création d'un projet et gestion des collaborateurs
![image](https://cloud.githubusercontent.com/assets/6525895/7086179/2fc401bc-df7e-11e4-949f-ea100e61036d.png)

La version avancée
![image](https://cloud.githubusercontent.com/assets/6525895/7086206/6a7f18f0-df7e-11e4-91cf-038e19030480.png)

- Gestion des journaux de bord
![image](https://cloud.githubusercontent.com/assets/6525895/7086220/80d646be-df7e-11e4-8843-5329b7b85305.png)

**Comment by valentinlefevre** [comment]
Réalisé avec @ChangruLU

Voilà les mockups Balsa liées aux imprimés écrans situés juste au dessus : 
![partager memo 1](https://cloud.githubusercontent.com/assets/6525895/7200709/eb300604-e4ff-11e4-8488-398bcec5c34d.png)
 Nous avons incorporé l'interface de rédaction d'un mémo avec celle de monitoring et ajouter un bouton de partage.

Au clic sur ce bouton nous arrivons sur cet affichage là :
![partager memo 2](https://cloud.githubusercontent.com/assets/6525895/7200744/2aeb27ba-e500-11e4-9eb1-9d1fa8dcdde7.png)
 Il faut se dire que tout le fond est grisé afin d'attirer l'attention de l'utilisateur sur les informations de la pop up permettant de gérer l'accessibilité du document par les collaborateurs et les personnes extérieures (professionnels, ...).

**Comment by Slaals** [comment]
Thansk @ChangruLU and @valentinlefevre for this work.

I would like to ** fix ** an important point relative to the issue #46. In your mock-up the user is consulting one specific memo, but the monitoring tool, that shows who consulted the memo or commented, is about all the memos.

As example, consider the memo titled "David", the list on the right would only lists the user(s) that consulted or commented the memo "David".

**Comment by PabloPrudhommeau** [comment]
Great job @valentinlefevre and @ChangruLU,

However, as you can see in the first position of the viewport on the functionnality "Rédiger un mémo" (https://github.com/Hypertopic/Cassandre/issues/40#issuecomment-91518124), there is the keywords highlighting functionnalities on the right side of the viewport, and in your mock-up, the activity is shown in the same place. As a result, we must reorganize these two crucial functionnalities for display each at different locations, isn't it ?

NB : @Slaals has begun a very nice and simplified mock-up (https://github.com/Hypertopic/Cassandre/issues/47#issuecomment-93969457) for explain the concept of viewport we should do to create a sequencing system organized step by step (Horizontal scrolling to move to the next step) .

**Comment by valentinlefevre** [comment]
@PabloPrudhommeau rien n'empêche d'envisager différents panneaux modulables (la techno le permet), je vais aménager le mockup pour que tu ais un aperçu plus clair de notre vision de tout ça.

**Comment by ConstantSIDJUI** [comment]
![11117391_860965990630701_1503054292_n](https://cloud.githubusercontent.com/assets/11837102/7226167/6106744c-e746-11e4-9c59-79cec49d213f.jpg)

L'image ci-dessus  présente les étapes d'un projet, et à chaque étape du projet on à la possibilité de partager le document et le document partagé contient une référence du document précédent.
La question qui se pose est la suivante: la personne à qui un document d'une étapes à été partagé, a-t-elle accès au document précédent (afin de savoir le contenu de l'étape précédente)? a-t-elle accès à tous les documents des différentes étapes?

**Comment by christophe-lejeune** [comment]
Ne mélangeons pas tout : le chaînage entre les différents types de comptes-rendus est traité ailleurs (entre autre #40 et #47). 

Il ressort de la discussion autour de la fonctionnalité #40 que l'analyste peut être amené à rédiger un compte-rendu de codage (en vert, étape 3) sur base d'un compte-rendu de terrain (en bleu, étape 1) sans étiqueter ce dernier (en rouge, étape 2). Cette étape n'est donc **pas** nécessaire. 

@ConstantSIDJUI, si un partage entraînait l'accès à tous les comptes-rendus liés, cela reviendrait, en définitive (par transitivité), à donner accès à tous le journal de bord, puisque les conclusions ramassent tout ce qui a été fait antérieurement. Si c'était ce que l'on voulait, nos précédents échanges seraient inutiles.

Il n'y a donc pas d'attribution de droits en cascade. 
* Seul le compte-rendu partagé doit effectivement être accessible.
* Les **références** des comptes-rendus sur lesquels s'appuie le compte-rendu partagés (ancrages) doivent être visibles. 
* Les comptes-rendus sur lesquels s'appuie le compte-rendu partagé ne sont, par contre, accessibles que s'ils sont partagés.

**Comment by christophe-lejeune** [comment]
* @Slaals a raison de préciser que l'affichage de la consultation n'a de sens que pour un compte-rendu déterminé.
* @valentinlefevre and @ChangruLU, l'historique des modifications des comptes-rendus peut **éventuellement** être également intégré. Mais pas nécessairement. Il faut être attentif à ce que la présente fonctionnalité ne fasse pas double emploi avec ce qui se dit à propos de la création d'un compte-rendu (#40)
* Au lu de ce que j'ai écrit par ailleurs (https://github.com/Hypertopic/Cassandre/issues/45#issuecomment-94307438), la deuxième maquette Balsamiq ne prévoit-elle pas trop de possibilités différentes ? Soit il s'agit d'une invitation (email), et la lecture entraîne nécessairement la possibilité de saisir un feedback. Soit il s'agit d'un partage avec un utilisateur (identifiant), qui peut, par définition rédiger un nouveau compte-rendu. La possibilité de modifier un compte-rendu est, pour sa part gérée via le champ auteur (#40) . 

À bien y réfléchir, je me demande s'il ne serait pas plus simple, en terme d'interface, de prévoir dans l'entête des comptes-rendus, un champ "lecteur(s)" sous le champ auteur(s). Les lecteurs invités pourraient apparaître en gris; ceux qui ont déjà consulté le compte-rendu, en noir.

**Comment by ConstantSIDJUI** [comment]
@christophe-lejeune la figure précédente précise que l'affichage des mémos se fait toujours en parellèle (affichage par deux: le mémo et ceux sur quoi il est ancré). de ce fait si on partage un mémo, il faut savoir ceux sur quoi il est ancré afin de le comprendre. 
Faut-il cacher l'ancrage ou l'autoriser?

**Comment by christophe-lejeune** [comment]
J'aurais tendance à maintenir ma proposition de ce matin : 
* Lorsqu'un compte-rendu est partagé, lui seul est effectivement être accessible.
* Les **ancrages** du compte-rendu partagé doivent être visibles dans son entête.

**Comment by valentinlefevre** [comment]
Comme nous l'avons envisagé avec @ChangruLU et proposé plus haut il nous apparaît intéressant de coller au fonctionnement actuel tout en intégrant les facilités de partage inhérentes aux technos utilisées (passage du crayon/papier au clavier/écran). 

Pour faire simple un projet est représenté par un dossier auquel contribue un certain nombre de collaborateurs conviés par le responsable du projet, chaque collaborateur se voit attribuer un journal de bord dont les règles seraient les suivantes :
- 1 journal de bord par collaborateur, 
- le contenu de chaque journal de bord est par défaut accessible en lecture et commentaire à tous les autres collaborateurs, 
- chaque collaborateur peut modifier les paramètres de partage de chacun de ses mémos (entrées du journal de bord), les rendre privés, modifiables par les autres collaborateurs, accessibles en lecture et commentaire à des personnes extérieures.

Nous avons envisagé ces paramètres afin de permettre un accès simplifié aux données relatives à un projet par tous ses collaborateurs tout en permettant à chacun de rédiger des éléments privés (malgré le fait que l'intérêt de l'outil soit de travail en collaboration, le travail chacun dans son coin étant, au vue de ses objectifs, est contre productif) et d'offrir à des personnes extérieures à l'équipe la possibilité d'apporter leur contribution par l'intermédiaire de commentaires.

Est ce que cela vous semble pertinent?

**Comment by bj0rge** [comment]
# Synthèse sur le partage

## Objectif

Avec le nombre grandissant de commentaires sur cette issue, il est nécessaire de procéder à une synthèse. Le but est :

- De repartir sur une base saine et claire pour tous.
- D'obtenir une validation par @christophe-lejeune afin de s'assurer que nous travaillons dans la bonne direction.

La synthèse est rédigée en français, toujours dans le même souci de clarté :)

## Partager un compte-rendu

- Dans un premier temps, il faut noter que chaque collaborateur au projet a par défaut un accès complet au dossier, et donc aux différents compte-rendus. Le partage ne concerne donc que des utilisateurs extérieurs (typiquement, la personne interviewée par exemple).
- Lorsqu'un compte-rendu est partagé, lui seul est accessible en lecture.
	- Les ancrages du CR sont visibles,
	- Les documents attachés par ancrage ne sont pas accessibles s'ils ne sont pas également partagés (*pas d'attribution de droits en cascade*).
- Le partage peut s'effectuer de deux façons différentes :
	- En autorisant un collaborateur déjà inscrit sur la plateforme à lire / commenter / éditer...
	- En créant une URL accessible de l'extérieur sans nécessité de création de compte (*plus souple pour un interviewé qui ne reviendra plus sur la plateforme par exemple*)
- Concernant l'historique des modification d'un CR, il faudrait décider si oui ou non il sera accessible lors d'un partage. Plusieurs solutions s'offrent à nous :
	- Un partage extérieur **ne donne jamais** accès aux versions précédentes. Nous partons alors du principe que ce qui est partagé est la version finale et que les versions successives (*brouillons en quelques sorte*) sont privées. Selon nous, c'est cette solution qui parait la plus logique (*lorsqu'on va à la boulangerie acheter du pain, il n'est pas nécessaire de savoir par quelles étapes il est passé pour savoir s'il est bon :)*).
	- Un partage extérieur **donne systématiquement** accès aux versions précédentes. Ici on considère donc qu'elles sont nécessaires à la bonne compréhension du CR.
	- Lors d'un partage extérieur, **il faut choisir** si on partage également l'historique. Cela risque de demander plus de développement pour une utilité peut-être relative, nous avons donc besoin de votre avis @christophe-lejeune : est-ce nécessaire ?
- Pour reprendre ce que @valentinlefevre a expliqué dans son commentaire précédent : **est-il nécessaire d'avoir un journal de bord privé par collaborateur** ? De ce que nous en comprenons, @valentinlefevre propose d'avoir un journal **qui ne serait pas accessible en écriture** aux autres membres du projet, afin d'y inscrire ses observations personnelles destinées à ne pas être éditées par quelqu'un d'autre, mais juste pouvoir être lu et commenté. @christophe-lejeune : faut-il se consacrer à cette fonctionnalité ?

Nous attendons vos réponses avant d'aller plus avant dans l'issue.

> Rédigé avec @ChangruLU

**Comment by christophe-lejeune** [comment]
Merci pour cette excellente synthèse !
* Je ne vois rien à redire sur les trois premiers points. Ils peuvent être validés sans réserve. 
* Comme @ChangruLU et @bj0rge, je ne vois pas de raison de donner accès à l'historique du compte-rendu partagé (le contraire pourrait d'ailleurs s'avérer gênant). La première option me semble s'imposer. 
* Il ne me semble pas nécessaire de créer une fonctionnalité dédiée aux journaux de bord individuels. En fait, à moins que  @valentinlefevre ait autre chose en tête, cette question recouvre deux situations bien distinctes, selon qu'un seul compte-rendu ou que le journal de bord entier soit privé. 
   * Au sein d'un projet mené à plusieurs, un des collaborateurs peut souhaiter se réserver l'accès à un compte-rendu particulier (un pense-bête personnel, une réflexion sur son propre positionnement). C'est bien un compte-rendu qui est rendu privé (et non un journal de bord entier). Ce cas de figure est vraisemblable et **doit être prévu**. 
   * Le cas particulier d'une recherche menée par un seul chercheur correspond à un "projet" ne comportant qu'un seul membre. A priori, ce qui aura été développé pour fonctionner à plusieurs fonctionnera également dans ce cas. Sauf erreur de ma part, ce cas particulier ne nécessite donc pas de développement spécifique. 

Les choses avancent bien. Voilà qui est stimulant. Je me réjouis de la suite !

**Comment by valentinlefevre** [comment]
Non aucune raison de créer une fonctionnalité particulière si vous n'en voyez pas l'absolue nécessitée.

**Comment by ChangruLU** [comment]
Test de recette pour le partage d'un memo avec RSpec syntax :

require 'memo'
feature 'Partager un memo' do
scenario 'Ajouter un collaborateur inscrit' do  
visit '/'
click_on 'Gérer le partage'
fill_in 'Partager avec', :with => 'christophe-lejeune'
select 'Droit', :with => 'Lire/Commenter/Modifier'
click_on 'Valider/Annuler'
visit '/'
expect(page).to have_content 'Memo partagé'
end

scenario 'Ajouter un intervenant extérieur' do  
visit '/'
click_on 'Gérer le partage'
fill_in 'Partager avec', :with => 'LU Changru'
select 'Droit', :with => 'Lire/Commenter'
click_on 'Valider/Annuler'
visit '/'
expect(page).to have_content 'Memo partagé'
end
end

Rédigé avec @valentinlefevre

**Comment by ConstantSIDJUI** [comment]
@ChangruLU dans le test de recette il manque l'URL d'accès au mémo au niveau du deuxième scénario : ajout d'un intervenant extérieur. Car s'agissant du partage avec l'interviewé, pour éviter qu'il crée de compte pour se loguer, une URL accessible de l'extérieur lui est envoyé.

**Comment by valentinlefevre** [comment]
@ConstantSIDJUI Les scénarii présentés plus haut sont vus du point de vue de l'utilisateur inscrit à l'outil et partageant le document en question. La partie du scénario présentant l'envoi automatisé d'un mail contenant l'URL d'accès au mémo pour l'intervenant extérieur ne nous apparaissait pas pertinente (dans le test de recette) car elle n'induit pas réellement d'action de l'utilisateur impliqué dans le scénario. La seule action à présenter étant un "click_on 'URL mémo'"...
